File: languages/ruby/lib/oso/oso/path_mapper.rb
# frozen_string_literal: true

module Oso
  module Oso
    # Map from a template string with capture groups of the form
    # `{name}` to a dictionary of the form `{name: captured_value}`
    class PathMapper
      def initialize(template:)
        capture_group = /({([^}]+)})/

        template = template.dup
        template.scan(capture_group).each do |outer, inner|
          template = if inner == '*'
                       template.gsub! outer, '.*'
                     else
                       template.gsub! outer, "(?<#{inner}>[^/]+)"
                     end
        end
        @pattern = /\A#{template}\Z/
      end

      def map(string)
        string.match(pattern)&.named_captures || {}
      end

      private

      attr_reader :pattern
    end
  end
end

File: languages/js/src/PathMapper.ts
import type { obj } from './types';

/**
 * Utility class to map from a forward-slash-delimited path to a dictionary
 * of matched path segments.
 */
export class PathMapper {
  #pattern: RegExp;

  constructor(template: string) {
    // TODO(gj): can probably do better than `[^}]` for the inner capture
    // group.
    const captureGroup = /({([^}]+)})/g;
    let temp = template;
    let captures;
    while ((captures = captureGroup.exec(template)) !== null) {
      const [, outer, inner] = captures;
      if (inner === '*') {
        temp = temp.replace(new RegExp(outer, 'g'), '.*');
      } else {
        temp = temp.replace(new RegExp(outer, 'g'), `(?<${inner}>[^/]+)`);
      }
    }
    this.#pattern = new RegExp(`^${temp}$`);
  }

  /**
   * Apply the templated pattern to a provided string, returning an object of
   * matching capture groups.
   */
  map(str: string): obj {
    return { ...this.#pattern.exec(str)?.groups };
  }
}
